Add logging to importer. Closes issue #192212

kyracho opened the pull request:
## Summary
Hello, this closes issue #192212 by adding a custom logger to the importer, making debugging issues easier.
### Checklist

Delete any items that are not applicable to this PR.

- [N/A] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [N/A] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios TODO
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed TODO
- [N/A] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [N/A] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [N/A] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [N/A] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [N/A] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### For maintainers

- [X]  This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process) (no API changes)


<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

kyracho commented:
Hello, I’ve now signed the contributor agreement, which covers the commit b523e16.

kyracho commented:
Hello, I wanted to kindly request that someone with the necessary permissions add the release note label to this pull request. Unfortunately, I don't have the authority to apply labels, and it seems the label is required to pass the CI test. Thanks in advance for your assistance.

TinaHeiligers commented:
@elasticmachine merge upstream

TinaHeiligers commented:
@kyracho thank you for your PR! I added the required labels and updated your branch.

TinaHeiligers commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

TinaHeiligers commented:
@kyracho the forced pushes don't start CI. Please request us to run CI on your behalf.

TinaHeiligers commented:
@elasticmachine merge upstream

elasticmachine commented:
There are no new commits on the base branch.

TinaHeiligers commented:
buildkite test this

kyracho commented:
@TinaHeiligers Thank you for the feedback! I’ve made the changes to pass the `logger` to the `SavedObjectsImporter` class and updated the `import` and `resolveImportErrors` APIs accordingly. I also (think) I made the necessary updates to the types, unit tests, and integration tests as you outlined. However, I haven’t had a chance to run the tests locally yet.

Regarding the force pushes—after rebasing onto main, some earlier commits were erased, which led me to force push. I’m still learning Git, so apologies for any confusion caused by that. Please let me know if you’d like me to make further adjustments or if anything else is needed!

TinaHeiligers commented:
> I’m fairly new to working on large codebases like this and with TypeScript in general, so I’m learning a lot as I go. I really appreciate your patience and any further guidance you can provide!

You chose one of the largest code bases to work on, congrats!
I've started a CI run and my team and I will help you fix any issues.

elasticmachine commented:
Pinging @elastic/kibana-core (Team:Core)

TinaHeiligers commented:
@kyracho you're probably facing the rebase nightmare by now. It would be a good idea to update your fork of Kibana and keep that up to date with your local clone. Don't try to rebase anymore, just merge updates.

TinaHeiligers commented on packages/core/saved-objects/core-saved-objects-import-export-server-internal/src/import/saved_objects_importer.ts:
```diff
@@ -57,40 +62,56 @@ export class SavedObjectsImporter implements ISavedObjectsImporter {
-    });
+    this.#log.debug('Starting the import process');
+    try {
+      const result = await importSavedObjectsFromStream({
```
Awaiting a resolved promise from `importSavedObjectsFromStream` is a problem when importing many saved objects and can cause Kibana to run out of memory. 
import shouldn't await on `importSavedObjectsFromStream`, we have to stick with calling the method directly.

I'd rather move the log line to before returning `importSavedObjectsFromStream` and revert the other changes here,

TinaHeiligers replied:
See [node streams](https://nodejs.org/api/stream.html#streams-promises-api) for more info

TinaHeiligers commented on packages/core/saved-objects/core-saved-objects-import-export-server-internal/src/import/saved_objects_importer.ts:
```diff
@@ -57,40 +62,56 @@ export class SavedObjectsImporter implements ISavedObjectsImporter {
   }
 
-  public resolveImportErrors({
+  public async resolveImportErrors({
```
we can't 'simply' change this....

TinaHeiligers commented on packages/core/saved-objects/core-saved-objects-import-export-server-internal/src/import/saved_objects_importer.ts:
```diff
@@ -57,40 +62,56 @@ export class SavedObjectsImporter implements ISavedObjectsImporter {
-    });
+    this.#log.debug('Resolving import errors');
+    try {
+      const result = await resolveSavedObjectsImportErrors({
```
similar comment as for `importSavedObjectsFromStream`

TinaHeiligers reviewed (changes requested):
import reads from a stream and awaiting the call (and other related function calls) can cause Kibana to OOM.

Logging doesn't warrant the risks of Kibana running out of memory.

TinaHeiligers reviewed (changes requested):
import reads from a stream and awaiting the call (and other related function calls) can cause Kibana to OOM.

Logging doesn't warrant the risks of Kibana running out of memory.

kyracho commented:
@TinaHeiligers  Thanks for the guidance! I’ve made the following changes in response to your comments:
- Removed `await` from `importSavedObjectsFromStream` and `resolveSavedObjectsImportErrors` to prevent memory issues.
- Moved the log statements to occur before the `return` statements, as you suggested
- Returned the promises directly
- Removed the `async` keywords from the methods

Let me know if anything else is needed!

TinaHeiligers commented:
@elasticmachine merge upstream

TinaHeiligers reviewed (approved):
This PR adds simplistic logging.
We need to enhance the logs with information on the saved objects being imported, how they are imported and if there are any errors. The enhancements can be a follow up to this PR.

TinaHeiligers commented:
buildkite test this

kyracho commented:
Thanks, Tina! I'd be happy to work on those enhancements in a new PR if that's okay. Please let me know if you'd like me to take it on.

TinaHeiligers commented:
buildkite test this

TinaHeiligers commented:
> Thanks, Tina! I'd be happy to work on those enhancements in a new PR if that's okay. Please let me know if you'd like me to take it on.

The idea is to use the logs for debugging, similar to how we log info during export.
Take a look at what we log for export with the following config to your `kibana.dev.yml`:
```//kibana.dev.yml
logging:
  appenders:
    so-console: // or use whatever you prefer
      type: console
      layout:
        type: pattern
        highlight: true
        pattern: "SOS_IMPEXP--[%date][%level][%logger]---%message"
  root:
    appenders: [console]
    level: info
  loggers:
    - name: savedobjects-service.exporter
      level: all
      appenders: [so-console] // feel free to change this
    - name: savedobjects-service.importer
      level: all
      appenders: [so-console] // feel free to change this
```
FYI: the appender is one I prefer to use, you can use whatever you want

Start es 
Start kibana in dev (any license is fine)
Once kibana's running, add any sample data. Adding sample data uses Saved Objects import (the logs you're adding here)
Navigate to the Saved Objects Management page and export anything from there (or all of them)
Watch the terminal you started Kibana from. The logs will be there as stdout

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/11227177901

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/240077)
* Commit: 30c29dccaca5f221c15013a55c4e7cf98ae3e0f8

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/240077","number":240077,"commit":"30c29dccaca5f221c15013a55c4e7cf98ae3e0f8"}],"number":240077}
buildkite-pr-comment-->

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/195339">](https://github.com/elastic/kibana/pull/195339)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
